[Species Balancing] Cyclorite Major Changes - Wave 1

Coolsurf6 opened the pull request:
## Short description
This PR requires:
Wave 2: Cyclorite proning nerf
https://github.com/ss14Starlight/space-station-14/pull/2719
Wave 3: Cyclorite medicine/poison nerf
https://github.com/ss14Starlight/space-station-14/pull/3089
This PR has information that relate to these other two PRs.

This PR stems from a suggestion post on the Starlight Discord about changes to the Cyclorite. Currently, the Cyclorite is an underused species with what seems to be a focus on increased health and durability, for lack of speed. This means that in most fights against other species and even mobs, Cyclorites will be unable to run away forcing them to stand their ground.

This is wave 1 of changes that I wish to make for the Cyclorite to give them more uniqueness and interesting playstyles. This is hopefully of a few waves of changes as some additions I have made should be reduced in favour of more active abilities that a lot of other species have such as Moths.

The current focus of Wave 1 is:
### Stat changes
Cyclorite resistances to **15% blunt** and **30% slash**. No burn or poison type weaknesses. - Completed :heavy_check_mark: 

Health increased to **150** (subject to decrease after active ability added) - Completed :heavy_check_mark: 

Damage threshold for slowdowns will be tweaked if needed to be more on the injured side. - Completed :heavy_check_mark: 

### Speed changes - To be worked on
Cyclorites now move 15% slower (from 10%) but now have a 50% reduction in anything that affects their movement speed in a **positive or negative way**. This means even Stimulents or Speedboots would have less of an effect on them, not just anything slowing them down.

The following movement speed systems are resisted by Cyclorites:
- Stamina Slowdown
- Contact Slowdown (Glue/Slime)
- Clothing Equipped
- Clothing/Item Held
- Wield-able Items (Hristrov)
- Pulling (A very minor amount really)
- Statuses (Like from Stimulents/Nocturine/Tear Gas and Magboots)
- Changeling 'StrainedMusclesActive' Speed
- Temperature slowdown from Cold
- Zombie (Rock Zombie.......)
- Ensnared Slowdown (Bolas I believe do this)
- Hunger (Changed due to scaling system having a bug)
- Thirst (Changed due to scaling system having a bug)

The main effects I refused to add to are:
- Damage Slowdown (Done manually through YAML)

### Mass changes - Completed :heavy_check_mark: 

Cyclorites now have an increased mass the same as a Diona making it easier for Cyclorites to drag heavy objects and people. Even though they are rock, they do have flesh internally. Didn't want to make them more than a Diona for now.

This also means dragging them will be slightly harder for most races without using items such as lockers, roller chairs etc. 

## Why we need to add this
I would like to give Cyclorites more unique interactions other than having slightly more health, night vision and a secondary back slot. Due to their decreased speed, they are mostly a detriment to play for a lot of roles such as engineering, medical for paramedics, and even cargo. Their increased mass never really affected them too much and due to their lack of speed, meant that being aggressive or trying to evade was especially difficult. 

Speed is king in SS14, and Cyclorites are by definition absolutely screwed over because of it. Instead of removing one of their unique tropes, I would rather lean in on it more. Let them be slow, but give people who try to use force significantly harder times to take them down. I still want to give chances for traitors and other antagonists a chance to contend especially though who know how to target a Cyclorite well with poisons ([WAVE 3](https://github.com/ss14Starlight/space-station-14/pull/3089)). 

I also believe that by making them very strong and durable, that healing them should take more time, not only due to them being able to walk into med bay with more wounds than a human could ever take and remain conscious, but due to their unique biology with their weaknesses to foreign chemicals like medicine and poisons, should be stuck in medical for longer and take more resources than usual forcing chemists to have to continue to top up more if the crew is high on Cyclorite casualties. ([WAVE 3](https://github.com/ss14Starlight/space-station-14/pull/3089))

## Other Ideas for different PRs
There are a lot more changes such as an ability that would give them more damage resistance similar to use with shields, which would need to be balanced with their health more but those are for other waves of changes.

_Another is to make Cyclorites slower to getting up taking double as much time to go from prone to standing, with potentially a different sound effect for when they go prone. This shouldn't affect Diona as I see them as more nimble while Cyclorites are more rigid even if they are the same weight._
Completed in [Wave 2](https://github.com/ss14Starlight/space-station-14/pull/2719).

## Media (Video/Screenshots)
https://github.com/user-attachments/assets/142e3a76-b15a-4f1f-97a1-f356a9ffc4bc
This shows the Slowness Resistance that causes Cyclorites to run roughly the same speed as a human in a security hardsuit, but also shows the slowdown from contact with the glue to be reduced more than the human.

<img width="1268" height="882" alt="image" src="https://github.com/user-attachments/assets/c94c12c9-d507-485b-83d4-d16ec71b8610" />
This is the updated Guidebook with all the changes listed nicely. If there are better ways to word it or things I have missed please let me know!

## Checks
<!-- check boxes for faster reviewing of your PR -->

- [x] I do not require assistance to complete the PR.
- [x] Before posting/requesting review of a PR, I have verified that the changes work.
- [x] I have added screenshots/videos of the changes, or this PR does not change in-game mechanics.
- [X] I affirm that my changes are licensed under the [Starlight Fork License](https://github.com/ss14Starlight/space-station-14/blob/Starlight/LICENSE-Starlight.TXT) and grant permission for use in this repository under its conditions.

**Changelog**
:cl: Coolsurf6
- tweak: Cyclorites walk at 15% slower speed (from 10%)
- tweak: Cyclorites now have 150 health and crit at 300
- tweak: Cyclorites resistances have been slightly increased
- tweak: Cyclorites are now as dense as Diona
- add: Cyclorites now slow down less from all sources but damage slowdown

Commit:
increased mass to 150% and decreased to 85% speed for cyclorites

Commit:
increased mass and increased health to 150

Commit:
added clarification on a todo

Commit:
actually commit the health changes to 150 this time

Commit:
added increased resistances to cyclorites

DocMoth commented:
To be clear at a fifteen percent slowdown, removing 75% percent of hardsuit slowdown and adding 50 health would make cyc insanely stronger than anyone else. A movement of 0.85 x (1 - 0.35(jug slowdown) x 0.25(reduction)) would equal 0.775625 compared to the 0.65 other jug users and this would be on top of cycs having 150 health and nightvision. Even cycs just wearing Security hardsuits would be better than any other secoff.

Commit:
fixed spelling mistake

Fasuh commented:
My stance is the same as i said on discord, i believe cyclorite should not get this high of resistance and health buffs, instead they should actually get the slowdown lowered, to make them faster in heavy suits (in my opinion it should be so they are about the same speed in a sec hardsuit and faster in jug and sec jug). 

The original idea had included also that they should be less affected by speed ups, like hyperzine or ephedrine, or speed boots, which would balance it out, you can be slow and tanky, better at holding heavy equipment, but you are stable, a felionoid will be probably way faster than you will ever be with hyperzine and their innate speed wearing a jugsuit.

Looking at it like "yeah but then they will be better" is just a flat thinking in my opinion, species are supposed to be better at certain things, this means that cyclorites would be way better with heavy gear, but then they would be worse with stuff like raidsuits, or BP vests due to their slowdown, maybe they would work better with riot suits, maybe there is some niche meta to stack slowing items, but thats what makes this interesting.

We cant balance stuff based on the top performing build, most of the time sec will not be wearing hardsuits, which means that cyclorites will be weaker in that time, but then they will slowly become better the moment its time to wear heavier equipment.

Commit:
added metabolizerScaleComponent and added to metaboliserSystem

Commit:
fixed narcoticscale spelling mistake

Commit:
added metabolizerscale to cyclorites with medicine inefficiency and poison/narcotic efficiency

Princess-Cheeseballs commented on Content.Server/Body/Systems/MetabolizerSystem.cs:
```diff
@@ -185,12 +186,31 @@ private void TryMetabolize(Entity<MetabolizerComponent, OrganComponent?, Solutio
+                        case "Narcotic":
+                            scaleRate = scaleComp.NarcoticScale;
+                            break;
+                    }
```
Instead of doing this you could include the data in the `group` above and just pass it.

Princess-Cheeseballs commented on Content.Server/Body/Systems/MetabolizerSystem.cs:
```diff
@@ -185,12 +186,31 @@ private void TryMetabolize(Entity<MetabolizerComponent, OrganComponent?, Solutio
                 mostToRemove = FixedPoint2.Clamp(rate, 0, quantity);
 
-                var scale = (float) mostToRemove;
+                var scale = (float) mostToRemove / scaleRate;
```
A lot safer and more efficient to use modifiers instead of divisors.

Arkanic replied:
using divisors is really sus, because small values create ridiculously large results

also, im REALLY sus of this math, because you are dividing by ScaleRate on a value that is *defined* initially to be proportional to scaleRate (`var rate = entry.MetabolismRate * group.MetabolismRateModifier * scaleRate;`)
its using a value twice, which would get simplified down to having no effect mathematically - need to double check what is going on here

also starlight comment

Princess-Cheeseballs committed:
ssafsasfa

asfafsa

Princess-Cheeseballs reviewed (changes requested):
Some suggestions, also made a PR for this branch to help with the other half of the PR.

Commit:
Merge pull request #1 from Princess-Cheeseballs/patch-1

Suggestions from DevBus

Commit:
added MovementSpeedModifierScaleComponent

Commit:
Merge remote-tracking branch 'Coolsurf6/cyclorite-changes-wave1' into cyclorite-changes-wave1

Commit:
added movementspeedmodifierscaleComponents and connected to multiple systems

Commit:
added MovementSpeedModifierScale to cyclorites with a 0.5 value

Commit:
added zombie speed being affected by scale

Commit:
added ensnared speed being affected by scale

Commit:
added contact speed being affected by scale

Commit:
decreased resistances values

Commit:
changed mass to be equal to diona at 300

Commit:
changed wording in comment

Coolsurf6 commented:
I would honestly love to have a scaling system as well for knockdowns so that Cyclorites get up twice as slow due to how rigid they are, but I might work on that in a later PR to keep things cleaner.

Coolsurf6 commented:
<img width="1466" height="513" alt="image" src="https://github.com/user-attachments/assets/0d61d9f7-335d-4d58-88f3-ce1b9188d700" />
This is the current speed values, with a 50% reduction rather than a 75%. I like that Security hardsuits being the lightest usually will be basically equal speed, but anything heavier actually has a difference. 

UPDATED:
<img width="1698" height="379" alt="image" src="https://github.com/user-attachments/assets/9821f87b-3769-4e60-aa9a-f45f2fe9b261" />

Coolsurf6 commented:
Uploaded videos of the major changes I believe should cover most. I have not addressed the review that Princess went over as I may need some more help with that for how best to clean it up.

Commit:
Merge branch 'starlight-dev' into cyclorite-changes-wave1

Coolsurf6 commented:
This is currently waiting on Maintainer feedback regarding changes, as well as the code quality.

Commit:
Merge branch 'starlight-dev' into cyclorite-changes-wave1

Commit:
increased health for SlowOnDamage using BaseMobSpecies as reference

Coolsurf6 commented:
Still requires updating the guidebook. Need to know if changes are suitable as guidebook additions are a little annoying to have to redo.

Arkanic commented on Content.Server/Body/Systems/MetabolizerSystem.cs:
```diff
@@ -185,12 +186,31 @@ private void TryMetabolize(Entity<MetabolizerComponent, OrganComponent?, Solutio
+                            break;
+                        case "Narcotic":
+                            scaleRate = scaleComp.NarcoticScale;
+                            break;
```
as metaboliser groups are dynamically created using yaml, the metabolizer scale component should support arbritrary groups, and not be hardcoded to anything

psuedo-ish code for this section:
```c#
if (TryComp<MetabolizerScaleComponent>(actualEntity, out var scaleComp) && scaleComp.MetabolizerScales.ContainsKey(group.Id))
    scaleRate = scaleComp.MetabolizerScales[group.Id];
```

rough idea of what you want to do as requested
Would also want to do the TryComp outside of the group foreach so that it isn't unecessarily trycomping N times

also, something you may have forgotten:
reagents are generally in *multiple* groups, not just one - with your current code, the reagent will just end up with the scale modifier that is the "last" in the list, you may want to *= so that all group modifiers get applied to the reagent - noting that if a chemical is both a narcotic and a poison it could create some unusually slow metab speeds

Coolsurf6 replied:
Metabolism Changes removed.

Arkanic commented on Content.Server/Body/Systems/MetabolizerSystem.cs:
```diff
@@ -185,12 +186,31 @@ private void TryMetabolize(Entity<MetabolizerComponent, OrganComponent?, Solutio
                 if (!proto.Metabolisms.TryGetValue(group.Id, out var entry))
                     continue;
 
-                var rate = entry.MetabolismRate * group.MetabolismRateModifier;
```
comment out this removal instead of deleting the line, so that it doesn't make it back in a later upstream merge

Coolsurf6 replied:
Metabolism Changes removed.

Commit:
updated cyclorite guidebook entry

Commit:
adjusted damage modifiers for cyclorites

Coolsurf6 commented:
I have gone ahead and not wanted to deal with any bullet damage DPS balancing at all, and instead removed the piercing and all heat related resistances.

Keeping it simple. 15% Blunt, 30% slash.

Arkanic commented on Content.Server/Body/Systems/MetabolizerSystem.cs:
```diff
@@ -185,12 +186,31 @@ private void TryMetabolize(Entity<MetabolizerComponent, OrganComponent?, Solutio
+                }
+                // Starlight-end
+                
+                var rate = entry.MetabolismRate * group.MetabolismRateModifier * scaleRate;
```
starlight comment

also, i think you should only be changing the HEAL/DAMAGE SCALE here, NOT consumption rate, as we already have mechanics for modifying reagent metabolisation rate (metabolizationRate on heart organ component, which is used for arachnids as well as vulpkanin)

at the moment both scale and rate is being affected by the same number, which is a bit opaque and weird behaviour, and ontop of that it's duplicating functionality that already exists

recommendation is to make it not affect rate, and only heal scale, and get the rate affect you want by changing the cyclorites heart organ yaml

Coolsurf6 replied:
Metabolism Changes removed.

Arkanic commented on Content.Server/_Starlight/Body/Components/MetabolizerScaleComponent.cs:
```diff
@@ -0,0 +1,28 @@
+    ///     Sets the scale for the "Medicine" reagent group efficiency and rate
+    /// </summary>
+    [DataField]
+    public float MedicineScale = 1f;
```
continuing from above, replace all of these individual floats with the following:

```c#
[DataField]
public Dictionary<ProtoId<MetabolismGroupPrototype>, float> MetabolizerScales = [];
```

Coolsurf6 replied:
Metabolism Changes removed.

Commit:
updated wording in cyclorite guidebook

Coolsurf6 commented:
Updated guidebook entry and added image to post.

Arkanic commented on Content.Shared/Movement/Systems/MovementSpeedModifierSystem.cs:
```diff
@@ -198,6 +206,18 @@ public void ModifySpeed(float mod)
         }
+
+        // Use these when you want the modifiers to be adjusted and pushed towards the center, aka for reagent and clothes modifiers.
+        public void ModifySpeedScaled(float walk, float sprint)
```
something really fishy is going on here, and i'll leave it to devs/maints to look into - changing how every speed modifier works fundamentally, without changing any values is concerning, but I don't fully understand the implications of these code changes - other than that this is
a) changing stuff WELL outside of the stated scope of the PR (cyclorite buffs, and this is fundamentally changing how ALL clothing speed modifiers work)
b) creates an incredibly large amount of changes on wizden files for no good reason, which is bound to create a lot of unecessary merge conflicts down the line (lines where you are replacing ModifySpeed with ModifySpeedScaled)

i know this is just for the cyclorites speed debuff/buff "scaling", but even so there is no reason for this to just change the actual movement function itself, instead of making a new function and renaming every instance of modifyspeed being used - there is no reason to have two versions of the method, because the scalar can just be like, 1f for all things which dont intentionally change it (which it should default to)

Arkanic commented on Content.Shared/Movement/Systems/MovementSpeedModifierSystem.cs:
```diff
@@ -181,7 +189,7 @@ public void ChangeBaseFriction(EntityUid uid, float friction, float frictionNoIn
     ///     call <see cref="MovementSpeedModifierSystem.RefreshMovementSpeedModifiers"/>.
     /// </summary>
-    public sealed class RefreshMovementSpeedModifiersEvent : EntityEventArgs, IInventoryRelayEvent
+    public sealed class RefreshMovementSpeedModifiersEvent(float scalar) : EntityEventArgs, IInventoryRelayEvent
```
starlight comment, not going to annotate all of them because there is a lot - everywhere that a wizden file is modified, you need to have a starlight comment indicating that it is our modification

it doesn't matter how tiny the change is, it is done so that when we do upstream merges we can see where we have changed code, so that our changes dont get overwritten during merge conflict resolution

Arkanic commented on Content.Shared/Movement/Systems/MovementSpeedModifierSystem.cs:
```diff
@@ -198,6 +206,18 @@ public void ModifySpeed(float mod)
+        // Use these when you want the modifiers to be adjusted and pushed towards the center, aka for reagent and clothes modifiers.
+        public void ModifySpeedScaled(float walk, float sprint)
+        {
+            WalkSpeedModifier *= ((walk - 1) * scalar + 1);
```
critical find,

can we double check this equation? the very first numbers i tried on my calculator i got a negative walk speed modifier out, which is not good
walk modifier of 0.5
scalar of 0.5 (50% less)
<img width="570" height="101" alt="image" src="https://github.com/user-attachments/assets/bcc2d996-dbf0-4dc8-a44d-9fb96b441039" />

also if you put in a walkspeed modifier of 1, it makes a multiply by zero which breaks everything

Coolsurf6 replied:
The equation I think you did was (walk - 1) * (scalar + 1). 

Scalar and + 1 are not ran first, Walk - 1 * scalar is and is then added to 1.

Personally I didn't make this equation.

Arkanic commented on Content.Shared/Movement/Systems/MovementSpeedModifierSystem.cs:
```diff
@@ -112,8 +113,15 @@ public void RefreshMovementSpeedModifiers(EntityUid uid, MovementSpeedModifierCo
+                scalar = comp.MovementSpeedScale;
+            }
+            
+            var ev = new RefreshMovementSpeedModifiersEvent(scalar);
```
i know *why* you are passing down the scalar to the event, because the function that does the proportional modification is in the event, which doesn't have access the ability to fetch components, but it's not a very good solution, and will likely need to change - potentially including moving the method outside of the event

Arkanic commented on Content.Shared/Movement/Systems/MovementSpeedModifierSystem.cs:
```diff
@@ -198,6 +206,18 @@ public void ModifySpeed(float mod)
+        public void ModifySpeedScaled(float walk, float sprint)
+        {
+            WalkSpeedModifier *= ((walk - 1) * scalar + 1);
+            SprintSpeedModifier *= ((sprint - 1) * scalar + 1);
```
also, why can't we just apply the scalar in the movement modifier system AFTER the event has fired, instead of applying it inside this event using some weird scalar pass down into the event?
you could get exactly the same number out of it, mathematically the answer is going to come out the same
and it would fix this massive mess of having two different methods, and changing a bunch of random upstream code method names

Arkanic commented on Content.Shared/Zombies/SharedZombieSystem.cs:
```diff
@@ -37,7 +37,7 @@ private void OnArmorExamine(Entity<ZombificationResistanceComponent> ent, ref Ar
     {
         var mod = component.ZombieMovementSpeedDebuff;
-        args.ModifySpeed(mod, mod);
+        args.ModifySpeedScaled(mod, mod);
```
same as all the stuff above, i strongly believe we don't need this changed method, and even if we did, we don't need a renamed method, and starlight comments required

Arkanic commented on Resources/Prototypes/_StarLight/Entities/Mobs/Species/cyclorites.yml:
```diff
@@ -73,6 +73,16 @@
       - ReagentId: CopperBlood
         Quantity: 1
+  - type: MetabolizerScale
+    medicineScale: 2
```
requested

something like:

```yaml
- type: MetabolizerScale
  metabolizerScales:
    Medicine: 2
    Posion: 0.5
    Narcotic: 0.5

Coolsurf6 replied:
Metabolism Changes removed.

Arkanic reviewed (changes requested):
not going to comment on balance, i think this is interesting and possible to make work, but the code does require some changes as it stands, particularly around the movement modifier stuff

couple of critical things picked up, again mostly around the movement changes

Commit:
removed metabolism changes due to potential upstream changes to metabolismGroups

Coolsurf6 commented:
Due to https://github.com/space-wizards/space-station-14/pull/42172 being a potential upstream change. I have opted to fully remove metabolism changes to Cyclorites in favor of another approach. 

No point coding in a function that is about to be removed :trollface:

Commit:
refactored movement speed changes

Commit:
Merge remote-tracking branch 'origin/starlight-dev' into cyclorite-changes-wave1

Commit:
attempted event creation for MovementSpeedModifierScale

Commit:
removed before tag for event

Commit:
Merge remote-tracking branch 'origin/starlight-dev' into cyclorite-changes-wave1

Commit:
added scaleModifier done by Rinary

Commit:
added bypass to hunger and thirst

Rinary1 commented:
Marked as stale

Coolsurf6 commented:
Going to remove the bypass due to a bug that stopped the bypass from working. We will see how balance is from there. 

I want this finally done so I can try working on the harder part being Chem stuff.

Commit:
removed scale for thirst and hunger

Commit:
changed slowondamage to still slow more even with movementmodifierscale

Commit:
updated guidebook for hunger & thirst slowdown removal

Coolsurf6 commented:
@Rinary1 removed the scaling portion of the modifierscalesystem due to the bug where after an event that would be run after another being skipped, it would scale both making it pointless.

Updated guidebook entry. 

I have kept cyclorites being slowed down from damage as that was able to be done through YAML.

This should be ready as well as the nerf in wave 2. Wave 3 being chems being scaled will have to come later unless you advise something else.

Commit:
Merge branch 'starlight-dev' into cyclorite-changes-wave1

Coolsurf6 commented:
I think the integration test is fake - GasSpecificHeats_Agree

Commit:
Merge branch 'starlight-dev' into cyclorite-changes-wave1

Commit:
removed unused comments

Commit:
removed unused comments once more

Commit:
spacing removal auuuuugh

Commit:
updated guidebook

Commit:
updated guidebook more to remove narcotics mention

Commit:
added mention of double time to stand and push up stamina changes to guidebook

Rinary1 commented on Content.Shared/Movement/Systems/MovementSpeedModifierSystem.cs:
```diff
@@ -184,12 +193,12 @@ public void ChangeBaseFriction(EntityUid uid, float friction, float frictionNoIn
-        public float WalkSpeedModifier { get; private set; } = 1.0f;
-        public float SprintSpeedModifier { get; private set; } = 1.0f;
+        public float WalkSpeedModifier { get; set; } = 1.0f;
+        public float SprintSpeedModifier { get; set; } = 1.0f;
```
```suggestion
        public float WalkSpeedModifier { get; private set; } = 1.0f;
        public float SprintSpeedModifier { get; private set; } = 1.0f;
```

Rinary1 commented on Content.Shared/Movement/Systems/MovementSpeedModifierSystem.cs:
```diff
@@ -184,12 +193,12 @@ public void ChangeBaseFriction(EntityUid uid, float friction, float frictionNoIn
     ///     call <see cref="MovementSpeedModifierSystem.RefreshMovementSpeedModifiers"/>.
     /// </summary>
-    public sealed class RefreshMovementSpeedModifiersEvent : EntityEventArgs, IInventoryRelayEvent
+    public sealed class RefreshMovementSpeedModifiersEvent: EntityEventArgs, IInventoryRelayEvent
```
```suggestion
    public sealed class RefreshMovementSpeedModifiersEvent : EntityEventArgs, IInventoryRelayEvent
```

Commit:
fixed spacing and readded private set

Coolsurf6 commented:
@Rinary1 reviews updated now. I don't fully remember why I had to remove the private set but could have been with one of the older ways of doing it that required that.

Seems to still be working fine with it set back to private.

Commit:
Update Cyclorite.xml via evil webedit to remove narcotics mention

Coolsurf6 commented:
Don't hurt me for web editing that small change. I only just saw it now :(
